File: shooting-game-backend/src/main/java/com/logicbottle/shooting/game/models/Action.java
package com.logicbottle.shooting.game.models;

import org.json.simple.JSONObject;

public class Action {
  private ActionType actionType;
  private JSONObject data;

  public Action() {
    data = new JSONObject();
  }

  public ActionType getActionType() {
    return actionType;
  }

  public void setActionType(final ActionType actionType) {
    this.actionType = actionType;
  }

  public JSONObject getData() {
    return data;
  }

  public Action addData(final String field, final Object value) {
    this.data.put(field, value);
    return this;
  }
}

File: shooting-game-frontend/src/app/models/action.ts
import { ActionType } from "./action-type";

export class Action {
    actionType: ActionType;
    data: any;

    constructor(actionType: any) {
        this.actionType = +ActionType[actionType];
    }

    static fromJson(rawJson: any): Action {
        let action = new Action(rawJson.actionType);
        action.data = rawJson.data;
        return action;
    }
}